crates/text_edit/src/lib.rs

```diff
@@ -185,17 +208,28 @@ impl TextEditBuilder {
     }
 }
 
-fn assert_disjoint_or_equal(indels: &mut [Indel]) {
+fn assert_disjoint_or_equal(indels: &mut Vec<Indel>) {
     assert!(check_disjoint_and_sort(indels));
 }
-// FIXME: Remove the impl Bound here, it shouldn't be needed
-fn check_disjoint_and_sort(indels: &mut [impl std::borrow::Borrow<Indel>]) -> bool {
-    indels.sort_by_key(|indel| (indel.borrow().delete.start(), indel.borrow().delete.end()));
-    indels.iter().zip(indels.iter().skip(1)).all(|(l, r)| {
-        let l = l.borrow();
-        let r = r.borrow();
-        l.delete.end() <= r.delete.start() || l == r
-    })
+
+fn check_disjoint_and_sort(indels: &mut Vec<Indel>) -> bool {
+    indels.sort_by_key(|indel| (indel.delete.start(), indel.delete.end()));
+    return check_disjoint(&mut indels.iter());
+}
+
+fn check_disjoint<'a, I>(indels: &mut I) -> bool
+where
+    I: std::iter::Iterator<Item = &'a Indel>,
+{
+    if let Some(indel) = indels.next() {
+        let mut indel = indel;
+        return indels.all(|right_indel| {
+            let check = indel <= right_indel || indel.eq_insert_at_offset(&right_indel);
+            indel = right_indel;
+            return check;
+        });
+    }
+    true
 }
```

Comment:
I am a bit confused about the rewrite here, the way you have written this now is quadratic over the number of edits which seems unnecessary. We know that the iterator is sorted already so doing the `indels.iter().zip(indels.iter().skip(1)).all` approach that was used before should still be applicable here

Reply:
I actually prefer the old way as well, but wasn't sure how to achieve it. The basic idea of this rewrite is so this check can run on an iterator, addressing that `fixme` about not having to allocate an intermediate vector. 

A technical question then, how can I achieve the old approach? Can I clone the iterator somehow? I can't just do 
`indels.zip(indels.skip(1)` sadly, and passing two iterators to the function seems a bit weird as well. Or am I wrong aabout the entire approach?

Sorry for being confusing, this particular PR is more about me learning to think rust, since I am noob. Feel free to point out obvious flaws, I am most possibly simply ignorant.

Reply:
> Sorry for being confusing, this particular PR is more about me learning to think rust, since I am noob.

No worries, you aren't the first one to contribute as a learning exercise., always happy to help.

Depending on what indels is you should be able to cheaply clone it. In the case here we can take a `indels: impl Iterator<Item = &'a Indel> + Clone` as the parameter, then you can `indels.clone().zip(indels.skip(1)`.

Reply:
Gotcha, makes sense!